### PR TITLE
Add MSBuildProjectBuilder

### PR DESCRIPTION
**Pull request by ryantrem**
This is an initial implementation of an MSBuild project builder. It includes a custom asset that can reference and configure an MSBuild project (`MSBuildProjectReference`), a custom asset importer (creates `MSBuildProjectReference`s from .csproj/.sln files), and logic for manually and automatically building referenced MSBuild projects (`MSBuildProjectBuilder`).

See https://github.com/microsoft/MSBuildForUnity/blob/user/ryantrem/MSBuildProjectBuilder/README.md for more details.

### CODE REVIEW

**Comment by jwittner** [comment]
Eek! Very exciting!

File: Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+
+namespace MSBuildForUnity
```
Comment (jwittner) [inline Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs]: Maybe `Microsoft.MSBuild` to better meet the [naming guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-namespaces) for namespaces?

File: Source/MSBuildProjectBuilder/Assets/Samples/OutputConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using UnityEngine;
+
+public class OutputConsumer : MonoBehaviour
```
Comment (jwittner) [inline Source/MSBuildProjectBuilder/Assets/Samples/OutputConsumer.cs]: Nit: Maybe put it in an examples namespace?

File: Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs (hunk 1 above)
Comment (ryantrem) [reply Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs]: MSBuild itself uses the `Microsoft.Build` namespace. Maybe the namespace could be `Microsoft.Build.Unity`?

Comment (jwittner) [reply Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs]: I generally dislike putting the platform into the namespace, but I guess in this case it's also the Feature? I can't think of a better option so I'm down. =)

Comment (jwittner) [reply Source/MSBuildProjectBuilder/Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilderEditor.cs]: I'd be fine with just `Microsoft.Build` too.

**Comment by jwittner** [comment]
Nit: When I create an `MSBuildProjectReferenc`e I can build/rebuild it and get the below esoteric error. Perhaps we want to disable the build buttons until something buildable is actually referenced? Or perhaps log an error about needing to specify one?

```
ArgumentException: Invalid path
System.IO.Path.GetDirectoryName (System.String path) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
MSBuildForUnity.MSBuildProjectBuilder+<RunDotNetAsync>d__15.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:222)
--- End of stack trace from previous location where exception was thrown ---
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
MSBuildForUnity.MSBuildProjectBuilder+<BuildProjectsAsync>d__9.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:149)
--- End of stack trace from previous location where exception was thrown ---
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
MSBuildForUnity.MSBuildProjectBuilder+<BuildProjectsAsync>d__8.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:108)
--- End of stack trace from previous location where exception was thrown ---
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
MSBuildForUnity.MSBuildProjectReference+MSBuildProjectReferenceEditor+<OnInspectorGUI>d__1.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectReferenceEditor.cs:24)
--- End of stack trace from previous location where exception was thrown ---
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at C:/buildslave/unity/build/Runtime/Export/UnitySynchronizationContext.cs:115)
UnityEngine.UnitySynchronizationContext:ExecuteTasks()
```

**Comment by jwittner** [comment]
Nit: Since a solution dropped in seems to have `AutoBuild` forced to `true` I think it'd make more sense to default new `MSBuildProjectReference` AutoBuild properties to `true` as well.

**Comment by jwittner** [comment]
Things seem to be working rather well! Really cool! Can't wait to pull this into our projects!

**Comment by ryantrem** [comment]
> Nit: When I create an MSBuildProjectReference I can build/rebuild it and get the below esoteric error. Perhaps we want to disable the build buttons until something buildable is actually referenced? Or perhaps log an error about needing to specify one?
> ArgumentException: Invalid path
> System.IO.Path.GetDirectoryName (System.String path) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> MSBuildForUnity.MSBuildProjectBuilder+<RunDotNetAsync>d__15.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:222)
> --- End of stack trace from previous location where exception was thrown ---
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> MSBuildForUnity.MSBuildProjectBuilder+<BuildProjectsAsync>d__9.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:149)
> --- End of stack trace from previous location where exception was thrown ---
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> MSBuildForUnity.MSBuildProjectBuilder+<BuildProjectsAsync>d__8.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectBuilder.cs:108)
> --- End of stack trace from previous location where exception was thrown ---
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> MSBuildForUnity.MSBuildProjectReference+MSBuildProjectReferenceEditor+<OnInspectorGUI>d__1.MoveNext () (at Assets/MSBuildProjectBuilder/Editor/MSBuildProjectReferenceEditor.cs:24)
> --- End of stack trace from previous location where exception was thrown ---
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) (at <df7127ba07dc446d9f5831a0ec7b1d63>:0)
> UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at C:/buildslave/unity/build/Runtime/Export/UnitySynchronizationContext.cs:115)
> UnityEngine.UnitySynchronizationContext:ExecuteTasks()

Updated, now we show an error instead of build/rebuild buttons in this case:
image

**Pull request merged**